### PR TITLE
Mech2d Util -> master

### PR DESCRIPTION
**Pull request by drorta**
Added a Mech2d util

### CODE REVIEW

File: src/main/java/frc/utils/mech2d/Mechanism2dUtil.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj.util.Color8Bit;
+
+public class Mechanism2dUtil {
+
+	private final Mechanism2d mechanism2d;
+	private MechanismLigament2d[] ligaments = new MechanismLigament2d[100];
+	private int numLigaments;
+
+	public Mechanism2dUtil(Mechanism2d mech2d) {
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dUtil.java:16]:
```suggestion
	public Mechanism2dUtil(Mechanism2d mechanism2d) {
```

File: src/main/java/frc/utils/mech2d/Mechanism2dUtil.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		ligaments[numLigaments] = ligament2d;
+		numLigaments++;
+	}
+
+	public void extendLigament(String ligament, String name, double length, Rotation2d angle, double width, Color8Bit color) {
+		MechanismLigament2d ligament2d = new MechanismLigament2d(name, length, angle.getDegrees(), width, color);
+		getLigament(ligament).append(ligament2d);
+		ligaments[numLigaments] = ligament2d;
+		numLigaments++;
+	}
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dUtil.java:54]: do DI, get the ligament not the things to build it

File: src/main/java/frc/utils/mech2d/Mechanism2dUtil.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		return getLigament(ligament).getColor();
+	}
+
+	public void setLigamentWidth(String ligament, double width) {
+		getLigament(ligament).setLineWeight(width);
+	}
+
+	public double getLigamentWidth(String ligament) {
+		return getLigament(ligament).getLineWeight();
+	}
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dUtil.java:100]: delete

File: src/main/java/frc/utils/mech2d/Mechanism2dUtil.java (hunk 1 above)
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dUtil.java:16]: get name

File: src/main/java/frc/utils/mech2d/Mechanism2dUtil.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+	public void publish() {
+		SmartDashboard.putData("Mechanism2d", mechanism2d);
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dUtil.java:103]: use name

File: src/main/java/frc/utils/mech2d/Mechanism2dUtil.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
+package frc.utils.mech2d;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.wpilibj.smartdashboard.Mechanism2d;
+import edu.wpi.first.wpilibj.smartdashboard.MechanismLigament2d;
+import edu.wpi.first.wpilibj.smartdashboard.MechanismRoot2d;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj.util.Color8Bit;
+
+public class Mechanism2dUtil {
```
Comment (hillelv) [inline src/main/java/frc/utils/mech2d/Mechanism2dUtil.java:10]:
```suggestion
public class Mechanism2dContainer {
```
consider

File: src/main/java/frc/utils/mech2d/Mechanism2dContainer.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import edu.wpi.first.wpilibj.smartdashboard.MechanismRoot2d;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj.util.Color8Bit;
+
+import java.util.HashMap;
+
+public class Mechanism2dContainer {
+
+	private final Mechanism2d mechanism2d;
+	private HashMap<String , MechanismLigament2d> ligaments = new HashMap<String, MechanismLigament2d>();
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dContainer.java:15]: final

Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dContainer.java:15]: initialize it in constructor

File: src/main/java/frc/utils/mech2d/Mechanism2dContainer.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		return ligaments.get(ligament).getColor();
+	}
+
+	public void setLigamentWidth(String ligament, double width) {
+		ligaments.get(ligament).setLineWeight(width);
+	}
+
+	public double getLigamentWidth(String ligament) {
+		return ligaments.get(ligament).getLineWeight();
+	}
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dContainer.java:76]: remove

File: src/main/java/frc/utils/mech2d/Mechanism2dContainer.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+		return ligaments.get(ligament).getLength();
+	}
+
+	public void setLigamentColor(String ligament, Color8Bit color) {
+		ligaments.get(ligament).setColor(color);
+	}
+
+	public Color8Bit getLigamentColor(String ligament) {
+		return ligaments.get(ligament).getColor();
+	}
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dContainer.java:68]: remove

File: src/main/java/frc/utils/mech2d/Mechanism2dContainer.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	public Rotation2d getLigamentAngle(String ligament) {
+		return Rotation2d.fromDegrees(ligaments.get(ligament).getAngle());
+	}
+
+	public void setLigamentLength(String ligament, double length) {
+		ligaments.get(ligament).setLength(length);
+	}
+
+	public double getLigamentLength(String ligament) {
+		return ligaments.get(ligament).getLength();
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dContainer.java:59]: remove

File: src/main/java/frc/utils/mech2d/Mechanism2dContainer.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+	public void setRootPosition(String root, double x, double y) {
+		getExistingRoot(root).setPosition(x, y);
+	}
+
+	public void setLigamentAngle(String ligament, Rotation2d angle) {
+		ligaments.get(ligament).setAngle(angle);
+	}
+
+	public Rotation2d getLigamentAngle(String ligament) {
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dContainer.java:50]: remove

File: src/main/java/frc/utils/mech2d/Mechanism2dContainer.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+	public void addLigament(String root, MechanismLigament2d ligament) {
+		getExistingRoot(root).append(ligament);
+		ligaments.put(ligament.getName(), ligament);
+	}
+
+	public void extendLigament(String addToLigament, MechanismLigament2d ligment) {
+		ligaments.get(addToLigament).append(ligment);
+		ligaments.put(ligment.getName(), ligment);
+
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dContainer.java:39]: remvoe line space

**Review by YHerm — CHANGES_REQUESTED** [review]

File: src/main/java/frc/utils/mech2d/Mechanism2dContainer.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj.util.Color8Bit;
+
+import java.util.HashMap;
+
+public class Mechanism2dContainer {
+
+	private final Mechanism2d mechanism2d;
+	private HashMap<String , MechanismLigament2d> ligaments = new HashMap<String, MechanismLigament2d>();
+	private String name;
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dContainer.java:16]: final

File: src/main/java/frc/utils/mech2d/Mechanism2dContainer.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+		ligaments.get(ligament).setLength(length);
+	}
+
+	public void setLigamentColor(String ligament, Color8Bit color) {
+		ligaments.get(ligament).setColor(color);
+	}
+
+	public void setLigamentWidth(String ligament, double width) {
+		ligaments.get(ligament).setLineWeight(width);
+	}
```
Comment (YHerm) [inline src/main/java/frc/utils/mech2d/Mechanism2dContainer.java:60]: do getRoot and getLigament

**Pull request closed without merging**